src/lib/shmem/src/shmalloc_impl.rs

```diff
@@ -0,0 +1,739 @@
+//! This module implements a low-level, unsafe shared memory allocator that uses mmap'ed shared
+//! memory files as the backing store. The module is intended to be no-std so it can be used in
+//! Shadow's shim library, which must async-signal-safe.
+//!
+//! The allocator chains together chunks of shared memory and divvies out portions of each chunk
+//! using a first-fit strategy. The allocator also implements a freelist so that allocated blocks
+//! can be reused efficiently after free. The allocator design isn't good for general-purpose
+//! allocation, but should be OK when used for just a few types.
+//!
+//! This code is intended to be private; the `allocator` module is the public, safer-to-use
+//! front end.
+
+use crate::raw_syscall::*;
+use numtoa::NumToA;
+
+const PATH_MAX_NBYTES: usize = 255;
+pub(crate) const PATH_BUF_NBYTES: usize = PATH_MAX_NBYTES + 1;
+pub(crate) type PathBuf = [u8; PATH_BUF_NBYTES];
+
+use vasi::VirtualAddressSpaceIndependent;
+
+fn get_null_path_buf() -> PathBuf {
+    [0; PATH_BUF_NBYTES]
+}
+
+#[cfg(debug_assertions)]
+type CanaryBuf = [u8; 4];
+
+#[cfg(debug_assertions)]
+fn get_canary_buf() -> CanaryBuf {
+    [0xDE, 0xAD, 0xBE, 0xEF]
+}
+
+#[cfg(not(debug_assertions))]
+type CanaryBuf = [u8; 0];
+
+#[cfg(not(debug_assertions))]
+fn get_canary_buf() -> CanaryBuf {
+    []
+}
+
+trait Canary {
+    fn canary_init(&mut self);
+    fn canary_check(&self) -> bool;
+
+    #[cfg(debug_assertions)]
+    fn canary_assert(&self) {
+        assert!(self.canary_check());
+    }
+
+    #[cfg(not(debug_assertions))]
+    fn canary_assert(&self) {}
+}
+
+pub(crate) enum AllocError {
+    Clock,
+    Open,
+    FTruncate,
+    MMap,
+    MUnmap,
+    Unlink,
+    WrongAllocator,
+    Leak,
+}
+
+pub(crate) fn log_error(error: AllocError, errno: Option<i32>) {
```

Comment:
Probably better to use the `log` crate than write directly to `stderr` --- The `log` crate supports `no_std`, and we have a `no_std` logger in the shim.

Maybe even better to have the error type include the errno value, implement `Debug` for that type, and propagate errors up to the user of this crate to let them decide what to do with it.

Reply:
fixed